### PR TITLE
travis: check build with various compiler options

### PR DESCRIPTION
**Pull request by mrc0mmand**
In the past we occasionally stumbled upon a build issue which could be
reproduced only with specific optimization level or other compilation
option. Let's try to build the current revision with several most common
compiler options causing such issues to catch them early.

---

Let's try to stick with Fedora Rawhide, even though it was known to be pretty unstable, so we can build with the latest gcc/llvm stack. If this proves to be unmaintainable, maybe we could either fall back to the latest stable Fedora version or maybe use an Arch Linux container.

Also, adding similar jobs using a Debian container would be probably wise as well, to cover older gcc/clang versions. Let's see first how expensive the build check actually is, though.

### CODE REVIEW

**Comment by mrc0mmand** [comment]
@keszybz I gave https://github.com/systemd/systemd-centos-ci/issues/143 a try and it looks like there's going to be some tweaking needed to compile with `--werror -Db_ndebug=true` successfully:

```
../src/journal/journal-send.c: In function 'sd_journal_printv':
../src/journal/journal-send.c:99:21: error: variable 'rlen' set but not used [-Werror=unused-but-set-variable]
   99 |                 int rlen;
      |                     ^~~~
../src/journal/journal-send.c: In function 'sd_journal_printv_with_location':
../src/journal/journal-send.c:477:21: error: variable 'rlen' set but not used [-Werror=unused-but-set-variable]
  477 |                 int rlen;
      |                     ^~~~
cc1: all warnings being treated as errors
```

or

```
../src/core/dbus-manager.c:350:18: error: unused variable 'm' [-Werror,-Wunused-variable]
        Manager *m = userdata;
                 ^
1 error generated.
```

are just the first ones, I guess there's going to be at least a few more.

**Comment by mrc0mmand** [comment]
I guessed wrong, after fixing these three the compilation passes.

File: src/journal/journal-send.c [hunk 1]
```diff
@@ -96,7 +96,7 @@ _public_ int sd_journal_printv(int priority, const char *format, va_list ap) {
 
         /* Allocate large buffer to accommodate big message */
         if (len >= LINE_MAX) {
-                int rlen;
+                _unused_ int rlen;
                 buffer = alloca(len + 9);
                 memcpy(buffer, "MESSAGE=", 8);
                 rlen = vsnprintf(buffer + 8, len + 1, format, ap);
```
Comment (keszybz) [inline src/journal/journal-send.c:102]: I'd prefer
```C
assert_se(vsnprintf(buffer + 8, len + 1, format, ap) == len);
```

File: src/journal/journal-send.c [hunk 2]
```diff
@@ -474,7 +474,7 @@ _public_ int sd_journal_printv_with_location(int priority, const char *file, con
...
```
Comment (keszybz) [inline src/journal/journal-send.c:480]: Same here.

**Comment by evverx** [comment]
> so we can build with the latest gcc/llvm stack

It should be possible to install gcc-10 and clang-10 directly via https://apt.llvm.org/ and https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test without launching a full-fledged container.

By the way, have you tried matrices? I think apart from running those similar jobs in parallel they would make it easier to figure out what fails there.

**Comment by mrc0mmand** [comment]
> > so we can build with the latest gcc/llvm stack
> 
> It should be possible to install gcc-10 and clang-10 directly via https://apt.llvm.org/ and https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test without launching a full-fledged container.

Ah, thanks, didn't know about this. That would definitely reduce the amount of overhead and boilerplate.

> 
> By the way, have you tried matrices? I think apart from running those similar jobs in parallel they would make it easier to figure out what fails there.

I thought about them, but at the back of my mind I was still aware that we encountered some obstacles last time, so I didn't bother. Looking at the current documentation they seem to be more powerful now than I remember, so it might be worth to give them a shot once again.

**Pull request merged**